torch/csrc/autograd/utils/grad_layout_contract.h

```diff
@@ -37,6 +37,11 @@ inline bool obeys_layout_contract(const at::Tensor& grad, const at::Tensor& vari
       }
     }
     return true;
+  } else if (variable.is_nested()) {
```

Comment:
This function gets called in accumulate_grad.h [here](https://github.com/drisspg/pytorch/blob/1f05f2b152508870d4c695e13f68fb5c3dfb5854/torch/csrc/autograd/functions/accumulate_grad.h#L111).  Nested tensor can only store flatten contagious nested tensors. I think there is desire for this to change, however in its current state I think that semantically nested tensors would obey the the layout contract. However deatch op is not implemented currently for nested tensors.  Should it be or is this accumulate grad case  long tail enough for it not to be needed in this PR?

Reply:
So the reason why I returned False here is that nested tensor does not have an implementation of detatch. Detach appears to be a composite kernel and because of these: https://github.com/pytorch/pytorch/issues/79048
nested tensor currently isn't supported. If this function returned true for nested tensors then it would lead to this code path: https://github.com/drisspg/pytorch/blob/c5a4b7e03ea39f10c896f8be7fab2f5446c8979c/torch/csrc/autograd/functions/accumulate_grad.h#L118

Reply:
`_has_same_shape` is also a composite kernel, is that working as intended?

Reply:
This is a great point, I can confirm via notebook.. Should probably add some tests for _has_same_shape() but to be honest not entirely sure where in the tests directory

when I run `nt.detach()`
```  Python
NotImplementedError: Could not run 'aten::detach' with arguments from the 'NestedTensorCPU' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build).
```

Reply:
`detach` is registered as CompositeExplicitAutograd, but `_has_same_shape` is CompositeImplicitAutograd, could that be a factor?

You can probably add the tests in `test_torch.py`

Reply:
Bingo after making `_has_same_shape` composite explicit when trying to run _has_same_shape as method on a nested tensor I get the same error:
`Could not run 'aten::_has_same_shape' with arguments from the 'NestedTensorCPU' backend.`

Reply:
If the issue is detach, you might need to add support for that to NestedTensorImpl.h akin to https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/SparseCsrTensorImpl.h#L83-L158